src/uct/cuda/cuda_copy/cuda_copy_md.c

```diff
@@ -286,52 +499,137 @@ static size_t uct_cuda_copy_md_get_total_device_mem(CUdevice cuda_device)
     return 1; /* return 1 byte to avoid division by zero */
 }
 
+static ucs_status_t
+uct_cuda_copy_detect_vmm(void *address, unsigned *is_vmm,
+                         ucs_memory_type_t *vmm_mem_type)
+{
+    ucs_status_t status        = UCS_OK;
+#if HAVE_CUDA_FABRIC
+    CUmemAllocationProp prop   = {};
+    CUmemGenericAllocationHandle alloc_handle;
+#endif
+
+    *is_vmm = 0;
+
+#if HAVE_CUDA_FABRIC
+    /* Check if memory is allocated using VMM API and see if host memory needs
+     * to be treated as pinned device memory */
+    status = UCT_CUDADRV_FUNC_LOG_DEBUG(
+            cuMemRetainAllocationHandle(&alloc_handle, (void*)address));
+    if (status != UCS_OK) {
+        status = UCS_OK;
+        goto out;
+    }
+
+    *is_vmm = 1;
+    status  = UCT_CUDADRV_FUNC_LOG_DEBUG(
+            cuMemGetAllocationPropertiesFromHandle(&prop, alloc_handle));
+    if (status != UCS_OK) {
+        status = UCS_ERR_INVALID_ADDR;
+    }
+
+    if ((prop.location.type == CU_MEM_LOCATION_TYPE_HOST) ||
+        (prop.location.type == CU_MEM_LOCATION_TYPE_HOST_NUMA) ||
+        (prop.location.type == CU_MEM_LOCATION_TYPE_HOST_NUMA_CURRENT)) {
+        *vmm_mem_type = UCS_MEMORY_TYPE_HOST;
+    } else if (prop.location.type == CU_MEM_LOCATION_TYPE_DEVICE) {
+        *vmm_mem_type = UCS_MEMORY_TYPE_CUDA;
+    } else {
+        *vmm_mem_type = UCS_MEMORY_TYPE_UNKNOWN;
+    }
+
+    UCT_CUDADRV_FUNC_LOG_DEBUG(cuMemRelease(alloc_handle));
+out:
+#endif
+    return status;
+}
+
+static void uct_cuda_copy_sync_memops(const void *address)
+{
+#if HAVE_CUDA_FABRIC
+    unsigned flags;
+#else
+    unsigned value = 1;
+#endif
+
+#if HAVE_CUDA_FABRIC
+    /* Synchronize future DMA operations for all memory types */
+    UCT_CUDADRV_FUNC_LOG_WARN(cuCtxGetFlags(&flags));
```

Comment:
1. can we cache on the md if we ever set sync memops? instead of doing extra function call to cuCtxGetFlags
2. unite the two ifdef macros in this function